### PR TITLE
mon/AuthMonitor: provide command to rotate the key for a user credential

### PR DESCRIPTION
**Pull request by batrick**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: doc/rados/operations/user-management.rst [hunk 1]
```diff
@@ -744,6 +744,20 @@ You may also :ref:`Modify user capabilities<modify-user-capabilities>` directly
...
+------------
+
+To rotate the secret for an entity, use:
+
+.. prompt:: bash #
+
+    ceph auth rotate <entity>
+
+This avoids the need to delete and recreate the key when it is compromised,
+lost, or just scheduled for rotation.
```
Comment (anthonyeleven) [inline doc/rados/operations/user-management.rst:758]: s/just //

Comment (batrick) [reply doc/rados/operations/user-management.rst]: done

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test make check

File: qa/workunits/cephtool/test.sh [hunk 2]
```diff
@@ -609,6 +609,20 @@ function test_auth()
   ceph auth del client.xx
   expect_false ceph auth get client.xx
 
+  # test rotation
+  ceph auth get-or-create client.admin2 mon 'allow *'
+  ceph auth get client.admin2 >> keyring1
+  env CEPH_KEYRING=keyring1 ceph -n client.admin2 auth get client.admin2 >> keyring2
+  diff -au keyring1 keyring2
```
Comment (rzarzynski) [inline qa/workunits/cephtool/test.sh:616]: ACK.

File: qa/workunits/cephtool/test.sh [hunk 3]
```diff
@@ -609,6 +609,20 @@ function test_auth()
...
+  # rotate itself
+  env CEPH_KEYRING=keyring1 ceph -n client.admin2 auth rotate client.admin2 >> keyring3
+  expect_false diff -au keyring1 keyring3
```
Comment (rzarzynski) [inline qa/workunits/cephtool/test.sh:619]: Yeah, secret should be different...

File: qa/workunits/cephtool/test.sh [hunk 4]
```diff
@@ -609,6 +609,20 @@ function test_auth()
...
+  expect_false env CEPH_KEYRING=keyring1 ceph -n client.admin2 auth get client.admin2
```
Comment (rzarzynski) [inline qa/workunits/cephtool/test.sh:620]: ... so we shouldn't be able to operate using the old one.

File: qa/workunits/cephtool/test.sh [hunk 5]
```diff
@@ -609,6 +609,20 @@ function test_auth()
...
+  expect_true env CEPH_KEYRING=keyring3 ceph -n client.admin2 auth get client.admin2
+  expect_true ceph auth get client.admin2 >> keyring4
+  expect_true diff -au keyring3 keyring4
+  rm keyring[1234]
```
Comment (rzarzynski) [inline qa/workunits/cephtool/test.sh:624]: I would love to see a check ensuring the only thing that got changed is the secret itself; particularly that the caps are untouched.

Comment (batrick) [reply qa/workunits/cephtool/test.sh:624]:
```
diff --git a/qa/workunits/cephtool/test.sh b/qa/workunits/cephtool/test.sh
index d024ce6912d4..266f133c6037 100755
--- a/qa/workunits/cephtool/test.sh
+++ b/qa/workunits/cephtool/test.sh
@@ -613,14 +613,20 @@ function test_auth()
   ceph auth get-or-create client.admin2 mon 'allow *'
   ceph auth get client.admin2 >> keyring1
   env CEPH_KEYRING=keyring1 ceph -n client.admin2 auth get client.admin2 >> keyring2
-  diff -au keyring1 keyring2
+  # they are the same:
+  expect_true diff -au keyring1 keyring2
   # rotate itself
   env CEPH_KEYRING=keyring1 ceph -n client.admin2 auth rotate client.admin2 >> keyring3
-  expect_false diff -au keyring1 keyring3
+  # only the key has changed:
+  diff -au keyring1 keyring3 | grep -E '^[-+][^-+]' | expect_false grep -v key
+  # the key in keyring1 no longer works:
   expect_false env CEPH_KEYRING=keyring1 ceph -n client.admin2 auth get client.admin2
+  # the key in keyring3 should work:
   expect_true env CEPH_KEYRING=keyring3 ceph -n client.admin2 auth get client.admin2
+  # now verify the key from `auth get` matches what rotate produced:
   expect_true ceph auth get client.admin2 >> keyring4
   expect_true diff -au keyring3 keyring4
+  expect_true ceph auth rm client.admin2
   rm keyring[1234]
 
   # (almost) interactive mode
```

Comment (rzarzynski) [reply qa/workunits/cephtool/test.sh:630]: :+1: Thank you!

**Review by rzarzynski — APPROVED** [review]

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66609](https://tracker.ceph.com/issues/66609).

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/66609#note-1

**Pull request merged**